core/src/main/scala/com/nvidia/spark/rapids/tool/analysis/AppSQLPlanAnalyzer.scala

```diff
@@ -354,43 +411,47 @@ class AppSQLPlanAnalyzer(app: AppBase, appIndex: Int) extends AppAnalysisBase(ap
 
   /**
-   * Generate IO-related diagnostic metrics for the SQL plan:
-   * output rows, scan time, output batches, buffer time, shuffle write time, fetch wait time, GPU
-   * decode time.
-   * @return a sequence of IODiagnosticResult
+   * Generate IO-related diagnostic metrics for the SQL plan. Metrics include:
+   * - Output rows
+   * - Scan time
+   * - Output batches
+   * - Buffer time
+   * - Shuffle write time
+   * - Fetch wait time
+   * - GPU decode time
+   *
+   * @return A sequence of `IODiagnosticResult` objects containing diagnostic metrics.
    */
   def generateIODiagnosticAccums(): Seq[IODiagnosticResult] = {
     val zeroRecord = StatisticsMetrics.ZERO_RECORD
-    IODiagnosticMetrics.toSeq.flatMap { case ((sqlId, nodeId, nodeName, stageIds), sqlAccums) =>
+
+    // Transform the diagnostic metrics map into a sequence of results
+    IODiagnosticMetricsMap.toSeq.flatMap { case ((sqlId, nodeId, stageIds), sqlAccums) =>
+      // Process each stage ID and compute diagnostic results
       stageIds.split(",").filter(_.nonEmpty).map(_.toInt).flatMap { stageId =>
-        val stageDiagnosticInfo = HashMap.empty[String, StatisticsMetrics]
+        val nodeName = sqlAccums.head.nodeName
+        val stageTaskIds = getStageTaskIds(stageId)
+        // A mapping from metric name to its statistical results (min, median, max, sum)
+        val metricNameToStatistics = HashMap.empty[String, StatisticsMetrics]
 
+        // Iterate through each IO metric
         sqlAccums.foreach { sqlAccum =>
-          val stageTaskIds = app.taskManager.getAllTasksStageAttempt(stageId).map(_.taskId).toSet
-          val accumInfo = app.accumManager.accumInfoMap.getOrElse(sqlAccum.accumulatorId,
-            new AccumInfo(AccumMetaRef(0L, AccumNameRef(""))))
-          val taskUpatesSubset =
-            accumInfo.taskUpdatesMap.filterKeys(stageTaskIds.contains).values.toSeq.sorted
-          if (taskUpatesSubset.nonEmpty) {
-            val min = taskUpatesSubset.head
-            val max = taskUpatesSubset.last
-            val sum = taskUpatesSubset.sum
-            val median = if (taskUpatesSubset.size % 2 == 0) {
-              val mid = taskUpatesSubset.size / 2
-              (taskUpatesSubset(mid) + taskUpatesSubset(mid - 1)) / 2
-            } else {
-              taskUpatesSubset(taskUpatesSubset.size / 2)
-            }
-            val metricName = if (sqlAccum.name.contains(OUTPUT_ROWS_METRIC)) {
-              OUTPUT_ROWS_METRIC
-            } else {
-              sqlAccum.name
-            }
-            stageDiagnosticInfo(metricName) = StatisticsMetrics(min, median, max, sum)
+          val accumInfo = app.accumManager.accumInfoMap.getOrElse(
+            sqlAccum.accumulatorId,
+            new AccumInfo(AccumMetaRef(0L, AccumNameRef("")))
+          )
+          // Compute the metric's statistics (min, median, max, sum) for the given stage
+          val accumInfoStatistics = getAccumInfoStatisticsInStage(accumInfo, stageTaskIds)
```

Comment:
This block seems to be deconstructing the tuple returned from `getAccumInfoStatisticsInStage()` and creates an instance of StatisticsMetrics using the individual values (min, med, max, sum).

Can `getAccumInfoStatisticsInStage()` directly return an `Option[StatisticsMetrics]` ?

Reply:
`getAccumInfoStatisticsInStage` is also used in `generateStageLevelAccums`. I didn't want to create an extra object in this function. WDYT?

Reply:
A tuple `(Long, Long, Long, Long)` is also an object.

Reply:
refactored this part, please see new changes.